File: 3.x.x/crates-utopia/dns/src/message/domain.rs
use crate::error::{Error, Result};

/// Domain name codec. PHP `Utopia\DNS\Message\Domain`.
#[derive(Debug)]
pub struct Domain;

impl Domain {
    pub const MAX_LABEL_LEN: usize = 63;
    pub const MAX_LABELS: usize = 127;
    pub const MAX_DOMAIN_NAME_LEN: usize = 255;

    /// PHP `Domain::encode`.
    pub fn encode(name: &str) -> Result<Vec<u8>> {
        if name.is_empty() {
            return Ok(vec![0]);
        }
        if name.ends_with("..") {
            return Err(Error::invalid("Domain labels must not be empty"));
        }
        let trimmed = name.trim_end_matches('.');
        if trimmed.is_empty() {
            return Ok(vec![0]);
        }

        let labels: Vec<&str> = trimmed.split('.').collect();
        let label_count = labels.len();
        if label_count > Self::MAX_LABELS {
            return Err(Error::invalid(format!(
                "Domain has too many labels: {label_count}"
            )));
        }

        let mut encoded = Vec::new();
        let mut total_length = 0usize;
        for label in labels {
            if label.is_empty() {
                return Err(Error::invalid("Domain labels must not be empty"));
            }
            if label.contains('@') {
                return Err(Error::invalid("Domain label contains invalid characters"));
            }
            let label_length = label.len();
            if label_length > Self::MAX_LABEL_LEN {
                return Err(Error::invalid(format!("Label too long: {label}")));
            }
            encoded.push(u8::try_from(label_length).unwrap_or(u8::MAX));
            encoded.extend_from_slice(label.as_bytes());
            total_length += label_length + 1;
        }
        total_length += 1;
        if total_length > Self::MAX_DOMAIN_NAME_LEN {
            return Err(Error::invalid(format!(
                "Encoded domain exceeds maximum length of {} bytes",
                Self::MAX_DOMAIN_NAME_LEN
            )));
        }
        encoded.push(0);
        Ok(encoded)
    }

    /// PHP `Domain::decode`. Updates `offset` to the first byte after the name.
    pub fn decode(data: &[u8], offset: &mut usize) -> Result<String> {
        let mut labels: Vec<String> = Vec::new();
        let mut jumped = false;
        let mut pos = *offset;
        let data_length = data.len();
        let mut visited_pointers = std::collections::HashSet::new();
        let mut label_count = 0usize;

        loop {
            if pos >= data_length {
                return Err(Error::decoding(
                    "Unexpected end of data while decoding domain name",
                ));
            }
            let len = data[pos];
            if len == 0 {
                if !jumped {
                    *offset = pos + 1;
                }
                break;
            }

            if len & 0xC0 == 0xC0 {
                if pos + 1 >= data_length {
                    return Err(Error::decoding(
                        "Truncated compression pointer in domain name",
                    ));
                }
                let pointer = (usize::from(len & 0x3F) << 8) | usize::from(data[pos + 1]);
                if pointer >= pos {
                    return Err(Error::decoding(
                        "Compression pointer must reference earlier position in packet",
                    ));
                }
                if pointer >= data_length {
                    return Err(Error::decoding(
                        "Compression pointer out of bounds in domain name",
                    ));
                }
                if !visited_pointers.insert(pointer) {
                    return Err(Error::decoding(
                        "Compression pointer loop detected in domain name",
                    ));
                }
                if !jumped {
                    *offset = pos + 2;
                }
                pos = pointer;
                jumped = true;
                continue;
            }

            if len & 0xC0 != 0 {
                return Err(Error::decoding(
                    "Reserved label type encountered in domain name",
                ));
            }

            let label_len = usize::from(len);
            if pos + 1 + label_len > data_length {
                return Err(Error::decoding(
                    "Label length exceeds remaining data while decoding domain name",
                ));
            }
            let label_bytes = &data[pos + 1..pos + 1 + label_len];
            labels.push(String::from_utf8_lossy(label_bytes).into_owned());
            label_count += 1;
            pos += label_len + 1;
            if label_count > Self::MAX_LABELS {
                return Err(Error::decoding("Domain name exceeds maximum label count"));
            }
            if !jumped {
                *offset = pos;
            }
        }

        Ok(labels.join("."))
    }
}

File: packages/dns/src/Message/Domain.php
<?php

declare(strict_types=1);

namespace Utopia\DNS\Message;

use Utopia\DNS\Exception\Message\DecodingException;

final readonly class Domain
{
    public const int MAX_LABEL_LEN = 63;
    public const int MAX_LABELS = 127;
    public const int MAX_DOMAIN_NAME_LEN = 255;

    /**
     * Encode a domain name according to RFC 1035.
     */
    public static function encode(string $name): string
    {
        if ($name === '') {
            return "\x00";
        }

        if (str_ends_with($name, '..')) {
            throw new \InvalidArgumentException('Domain labels must not be empty');
        }

        $trimmed = rtrim($name, '.');
        if ($trimmed === '') {
            return "\x00";
        }

        $labels = explode('.', $trimmed);
        $labelCount = \count($labels);

        if ($labelCount > self::MAX_LABELS) {
            throw new \InvalidArgumentException("Domain has too many labels: $labelCount");
        }

        $encoded = '';
        $totalLength = 0;

        foreach ($labels as $label) {
            if ($label === '') {
                throw new \InvalidArgumentException('Domain labels must not be empty');
            }

            if (str_contains($label, '@')) {
                throw new \InvalidArgumentException('Domain label contains invalid characters');
            }

            $labelLength = \strlen($label);

            if ($labelLength > self::MAX_LABEL_LEN) {
                throw new \InvalidArgumentException("Label too long: $label");
            }

            $encoded .= \chr($labelLength) . $label;
            $totalLength += $labelLength + 1; // length byte + label
        }

        $totalLength += 1; // trailing zero-length octet

        if ($totalLength > self::MAX_DOMAIN_NAME_LEN) {
            throw new \InvalidArgumentException(
                'Encoded domain exceeds maximum length of ' . self::MAX_DOMAIN_NAME_LEN . ' bytes',
            );
        }

        return $encoded . "\x00";
    }

    /**
     * Decode a domain name from DNS wire format, handling compression pointers.
     *
     * Per RFC 1035 Section 4.1.4, compression pointers allow domain names to
     * reference earlier occurrences in the packet. This implementation tracks
     * visited pointer positions to prevent infinite loops from malicious packets.
     *
     * @param string $data   Full DNS packet
     * @param int    $offset Current read offset (updated to first byte after the name)
     * @return string Decoded domain name in dotted form
     *
     * @throws DecodingException when the packet is malformed or contains pointer loops.
     */
    public static function decode(string $data, int &$offset): string
    {
        $labels = [];
        $jumped = false;
        $pos = $offset;
        $dataLength = \strlen($data);

        // Track visited pointer positions to detect loops (RFC 1035 compliance)
        // This is more reliable than iteration counting as it catches actual cycles
        $visitedPointers = [];

        // Maximum labels per RFC 1035 (127 labels * 63 chars + separators = 255 max)
        $labelCount = 0;

        while (true) {
            if ($pos >= $dataLength) {
                throw new DecodingException(
                    'Unexpected end of data while decoding domain name',
                );
            }

            $len = \ord($data[$pos]);
            if ($len === 0) {
                if (!$jumped) {
                    $offset = $pos + 1;
                }
                break;
            }

            if (($len & 0xC0) === 0xC0) {
                if ($pos + 1 >= $dataLength) {
                    throw new DecodingException(
                        'Truncated compression pointer in domain name',
                    );
                }

                $pointer = (($len & 0x3F) << 8) | \ord($data[$pos + 1]);

                // RFC 1035: Pointer must reference earlier in packet (forward refs invalid)
                if ($pointer >= $pos) {
                    throw new DecodingException(
                        'Compression pointer must reference earlier position in packet',
                    );
                }

                if ($pointer >= $dataLength) {
                    throw new DecodingException(
                        'Compression pointer out of bounds in domain name',
                    );
                }

                // Detect pointer loops by tracking visited positions
                if (isset($visitedPointers[$pointer])) {
                    throw new DecodingException(
                        'Compression pointer loop detected in domain name',
                    );
                }
                $visitedPointers[$pointer] = true;

                if (!$jumped) {
                    $offset = $pos + 2;
                }
                $pos = $pointer;
                $jumped = true;
                continue;
            }

            // Check for reserved label type (RFC 1035: bits 6-7 indicate label type)
            // 00 = standard label, 11 = compression pointer, 01/10 = reserved
            if (($len & 0xC0) !== 0) {
                throw new DecodingException(
                    'Reserved label type encountered in domain name',
                );
            }

            if ($pos + 1 + $len > $dataLength) {
                throw new DecodingException(
                    'Label length exceeds remaining data while decoding domain name',
                );
            }

            $labels[] = substr($data, $pos + 1, $len);
            $labelCount++;
            $pos += $len + 1;

            if ($labelCount > self::MAX_LABELS) {
                throw new DecodingException(
                    'Domain name exceeds maximum label count',
                );
            }

            if (!$jumped) {
                $offset = $pos;
            }
        }

        return implode('.', $labels);
    }
}
